### ISSUE TITLE
Unit price is not rounded the same as product price?

### ISSUE THREAD

**Issue by Garfield2013** [PrestaShop/PrestaShop#28621]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

Hi there,

There seems to be a bug in regards to rounding of Unit-price: 

![image](https://user-images.githubusercontent.com/23308427/171623282-345b9beb-5f8a-4d80-b38d-f2f66dc00539.png)

![image](https://user-images.githubusercontent.com/23308427/171603799-15e023f0-0af6-4489-b851-3d23d83a57c8.png)

### Expected behavior

Expected: AED 600.00
Actual: AED 600.01

### Steps to reproduce

1. **BO**>**Catalog**>**Products**
2. Create a product and set price to: 163.49 USD
3. **BO**>**Shop Parameters**>**General**
4. Set `Round Mode` to: `Round down to the nearest value`
5. Set `Round Type` to: `Round on each item`
6. Save
7. **BO**>**International**>**Localization**>**Currencies**
8. Set manually the Conversion rate between USD/AED to 3.67
9. Go to FO > change **currency** to AED
10. Add the product in the cart and checkout
11. See error > price: AED 600.01 on the detail of product in your cart 
12. Clic on "Proceed to checkout"
13. Clic on the "show details" button on the right 
14. See error > price: AED 600.01

### PrestaShop version(s) where the bug happened

1.7.8.6 - 1.7.7.8 - 1.7.6.5 - develop

### PHP version(s) where the bug happened

7.4.29 - 7.3.33

**Comment by AureRita** [comment]
Hi @Garfield2013 

Thank you for the reporting Issue.
Currently I can't reproduce your issue, can you tell me the round mode that you've and the round type ? (you can see this on BO>Shop Parameters>General)
You can see that I get on this video : 

https://user-images.githubusercontent.com/100699445/171004204-5d5183bf-fa75-451a-a0d7-d2bde22d2da0.mp4

Thanks

**Comment by Garfield2013** [comment]
In your video one price of 600.01 always remains? I need the rounding to be
the same everywhere to 600.00

Btw. my price is set to 163.49 USD in the B.O. and conversion rate between
USD/AED is set manually to 3.67
The price of 600.00/600.01 in my setup is in AED currency (selected as
currency in the top of the page).

**Comment by hibatallahAouadni** [comment]
Hello @Garfield2013 

Could you please answer @AureRita question:
> can you tell me the round mode that you've and the round type ? (you can see this on BO>Shop Parameters>General)

Waiting for your feedback.

Thanks!

**Comment by Garfield2013** [comment]
![image](https://user-images.githubusercontent.com/23308427/171603716-784384c9-0c40-489c-8b80-04d099172fdd.png)

**Comment by hibatallahAouadni** [comment]
Hello @Garfield2013 

With all provided infos in your comments, I setted steps to reprouce the issue, could you please check it :pray: 
And following these steps, I didn't manage to reproduce this issue with PS1786 and develop, see the attached screen record below:

https://user-images.githubusercontent.com/23308427/171606753-7185ee1c-5ef5-467d-8c63-bf7b4b63de65.mp4


Please check and feedback.
Thanks!

**Comment by Garfield2013** [comment]
Yes you manage to get this price to 600.00 AED but it is not the issue. Look at the screenshots in my first post. Hint: click add to cart and then you will see 600.01 AED and also continue to the payment step (step 4) and check all the listed prices. Some er OK 600.00 and some are not 600.01

**Comment by AureRita** [comment]
Hi @Garfield2013 , 

Thank you for the time you pass on this issue, I reproduced your issue as we can see on this video : 

https://user-images.githubusercontent.com/100699445/171654354-30fb8b00-38e8-4739-9d67-378b3e1a4f91.mp4

I'll change the end of the step to reproduce to have all the details, could you checked and say me if it's correct

**Comment by Garfield2013** [comment]
I answered this the 2nd June 2022 (I don't know why my response is not showing up here)?

https://www.screencast.com/t/46m1yDQ6JhrG
https://www.screencast.com/t/g7YrHPLO4B

Please fix this as soon as possible and give me an update. Thank you so much!

**Comment by AureRita** [comment]
Hi @Garfield2013 ,

There is no problem, this issue is already in our issue database to be processed

Please be aware that some issues might take a very long time to be resolved. If this one is important to you and you cannot wait for it to be fixed on the project’s own time, we strongly suggest you consider [contacting a professional](https://www.prestashop-project.org/support/) to help you. 

Thanks

**Comment by Garfield2013** [comment]
You have already started to work on this issue and fixed 2 of 3 bugs - would it not be best for everybody to fix the final part now as well? :)

**Comment by Garfield2013** [comment]
You still need to fix this: https://www.screencast.com/t/bR1cF1DTjnuG

I guess that would fix the unit-price in the bottom of the payment-page as
well (check my screenshots in my first post).
Otherwise unit-price needs the same fix as well at the bottom of the
payment-page.

**Comment by RosaBenouamer** [comment]
I have change the label with @AureRita :-)

**Comment by RosaBenouamer** [comment]
As seen with @AureRita when correcting this bug check the both place 
![image](https://github.com/PrestaShop/PrestaShop/assets/124665010/0ac02980-ba5b-454b-a853-826afe0240c5)
and 
![image](https://github.com/PrestaShop/PrestaShop/assets/124665010/16a26e50-80f9-4d37-b737-42d0c79d59ee)

**Comment by boo-code** [comment]
Reproduced on 9.2.x with @Garfield2013's exact numbers, and the cause is one missing pair of roundings.

`163.49 * 3.67 = 600.0083`. With `Round down` and `Round on each item`:

```
cart row price / price_wt      600.0083      (raw, as Cart::getProducts returns them)
cart row total / total_wt      600.0
presented price                AED600.00
presented regular_price        AED600.01     <- the amount in the report
summary total_products         600
```

**`CartLazyArray` rounds two of the four per-unit values it passes on:**

```php
$rawProduct['price']    = Tools::ps_round($rawProduct['price'],    Context::getContext()->getComputingPrecision());
$rawProduct['price_wt'] = Tools::ps_round($rawProduct['price_wt'], Context::getContext()->getComputingPrecision());
```

`price_without_reduction` and `price_without_reduction_without_tax` go through untouched. `ProductLazyArray` then formats all four with the same `PriceFormatter`, which rounds with the locale's rule rather than the shop's — half up. So the selling price is rounded down to `600.00` before formatting and the price before reduction is rounded half up to `600.01` during it. Same line, same product, two different rounding rules.

That also explains why the total is right while the line is wrong, and why @AureRita's first video showed 600.01 everywhere until the round mode was set: the discrepancy only appears when the shop's rounding mode disagrees with half up on that particular fraction.

**Prepared** against `9.2.x`: the two values before reduction are rounded where the two selling prices already are.

```
                          before      after
presented price           AED600.00   AED600.00
presented regular_price   AED600.01   AED600.00
```

Covered by a scenario next to the existing rounding features, `Cart/Rounding/displayed_price_before_reduction.feature`, with a price carrying more decimals than the currency under `round down` and under `round up`. The new assertion compares the two displayed amounts to each other rather than to a literal, so it does not depend on the fixture's tax rate or currency. With the presenter reverted it fails as `Product "product1" is displayed at $600.00 but with a price before reduction of $600.01`, and the mirror under round up.

The full `cart` suite (244 scenarios, 3548 steps) and the full `order` suite (260 scenarios, 7344 steps) pass on the branch.

@RosaBenouamer, on your 2023 note about checking both places: the cart page and the checkout summary both read `regular_price` from this same presenter, so both follow from this one change.

The branch is ready; a PR can be opened on request.
